src/v/kafka/server/write_at_offset_stm.cc

```diff
@@ -160,15 +164,20 @@ ss::future<result<raft::replicate_result>> write_at_offset_stm::do_replicate(
           kafka::offset_cast(kafka::next_offset(stm_last_offset)),
           model::prev_offset(kafka::offset_cast(expected_base_offset)),
           model::term_id{0});
-        batches.reserve(ghost_batches.size() + 1);
-        std::ranges::move(ghost_batches, std::back_inserter(batches));
+        effective_batches.reserve(ghost_batches.size() + 1);
+        std::ranges::move(ghost_batches, std::back_inserter(effective_batches));
     }
 
+    auto batch_start_offset = expected_base_offset;
+    kafka::offset batch_end_offset;
+    for (auto& b : batches) {
```

Comment:
I think this is not correct, the signature of the replicate should be changed to

```
s::future<result<raft::replicate_result>> write_at_offset_stm::do_replicate(
  chunked_vector<model::record_batch> batches,
  chunked_vector<kafka::offset> expected_base_offsets,
```

to correctly fill in gaps (eg: from compaction), I'll make that change.